[SPARK-26196][SPARK-26281][WebUI] Total tasks title in the stage page is incorrect when there are failed or killed tasks and update duration metrics

shahidki31 opened the pull request:
## What changes were proposed in this pull request?
This PR fixes 3 issues
1) Total tasks message in the tasks table is incorrect, when there are failed or killed tasks
2) Sorting of the "Duration" column is not correct
3) Duration in the aggregated tasks summary table and the tasks table and not matching. 

Total tasks  = numCompleteTasks +  numActiveTasks + numKilledTasks + numFailedTasks;

Corrected the duration metrics in the tasks table as executorRunTime based on the PR https://github.com/apache/spark/pull/23081

## How was this patch tested?
test step:
1)
```
bin/spark-shell 
scala > sc.parallelize(1 to 100, 10).map{ x => throw new RuntimeException("Bad executor")}.collect()
```
![screenshot from 2018-11-28 07-26-00](https://user-images.githubusercontent.com/23054875/49123523-e2691880-f2de-11e8-9c16-60d1865e6e77.png)

After patch:
![screenshot from 2018-11-28 07-24-31](https://user-images.githubusercontent.com/23054875/49123525-e432dc00-f2de-11e8-89ca-4a53e19c9c18.png)


2)  Duration metrics:
Before patch:
![screenshot from 2018-12-06 03-25-14](https://user-images.githubusercontent.com/23054875/49546591-9e8d9900-f906-11e8-8a0b-157742c47655.png)

After patch:
![screenshot from 2018-12-06 03-23-14](https://user-images.githubusercontent.com/23054875/49546589-9cc3d580-f906-11e8-827f-52ef8ffdeaec.png)

shahidki31 committed:
Total tasks title in the stage page is incorrect

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5422/
Test PASSed.

SparkQA commented:
**[Test build #99345 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99345/testReport)** for PR 23160 at commit [`37bcd62`](https://github.com/apache/spark/commit/37bcd6231816c7bc2b2561bff10955b822934ac6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5423/
Test PASSed.

SparkQA commented:
**[Test build #99346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99346/testReport)** for PR 23160 at commit [`bbd745a`](https://github.com/apache/spark/commit/bbd745ac0f727de5988bcc876e57d11a32eadb31).

SparkQA commented:
**[Test build #99346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99346/testReport)** for PR 23160 at commit [`bbd745a`](https://github.com/apache/spark/commit/bbd745ac0f727de5988bcc876e57d11a32eadb31).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99346/
Test PASSed.

SparkQA commented:
**[Test build #99345 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99345/testReport)** for PR 23160 at commit [`37bcd62`](https://github.com/apache/spark/commit/37bcd6231816c7bc2b2561bff10955b822934ac6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

gatorsmile commented:
cc @gengliangwang

gengliangwang reviewed (approved):
LGTM. Please add tag [WebUI] in the PR title

gengliangwang commented:
cc @tgravescs @pgandhi999

tgravescs commented:
reviewing, to clarify @pgandhi999 question, I assume the screenshots above are cut off and the task table is really showing 7 total failed tasks?

shahidki31 commented:
@tgravescs Yes. Task table is showing all the tasks. Sorry, I didn't put the whole screen shot.

shahidki31 commented on core/src/main/resources/org/apache/spark/ui/static/stagepage.js:
```diff
@@ -610,7 +610,8 @@ $(document).ready(function () {
 
                 // building tasks table that uses server side functionality
-                var totalTasksToShow = responseBody.numCompleteTasks + responseBody.numActiveTasks;
+                var totalTasksToShow = responseBody.numCompleteTasks + responseBody.numActiveTasks +
```
Yes @pgandhi999 . All the failed tasks are displaying in the tasks table. Thanks for your great work.

tgravescs reviewed (approved):
+1, thanks @shahidki31 !

shahidki31 committed:
change duration as executorRunTime

SparkQA commented:
**[Test build #99542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99542/testReport)** for PR 23160 at commit [`71aff34`](https://github.com/apache/spark/commit/71aff3447b4c1a80e168a18202864191df189709).

shahidki31 commented:
Hi @pgandhi999 , There is a small nit there. Recently we fixed a bug related to the "duration" metrics in the tasks table (see https://github.com/apache/spark/pull/23081), but that hasn't reflected here. I have updated the code

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5604/
Test PASSed.

SparkQA commented:
**[Test build #99543 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99543/testReport)** for PR 23160 at commit [`661566b`](https://github.com/apache/spark/commit/661566bd51efb40176327a58e5bed5dcfe7cacf0).

pgandhi999 commented on core/src/main/resources/org/apache/spark/ui/static/stagepage.js:
```diff
@@ -661,8 +662,8 @@ $(document).ready(function () {
-                                if (row.duration) {
-                                    return type === 'display' ? formatDuration(row.duration) : row.duration;
+                                if (row.taskMetrics && row.taskMetrics.executorRunTime) {
+                                    return type === 'display' ? formatDuration(row.taskMetrics.executorRunTime) : row.taskMetrics.executorRunTime;
```
Ok, that makes sense. However, you also need to update the search filter method to not search on duration in StagesResource.scala here:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala#L213.

shahidki31 replied:
Thanks removed.

pgandhi999 reviewed (commented):
My guess is that this might break the sorting on the Duration column in the tasks table as we are using the value of the name field in the column data to match with the appropriate name of the data field in the server data below in StagesResource:

`withUI(_.store.taskList(stageId, stageAttemptId, pageStartIndex, pageLength,
      indexName(columnNameToSort), isAscendingStr.equalsIgnoreCase("asc")))
  }`
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/api/v1/StagesResource.scala#L178. 
So as far as I can think, I expect that if you sort the Duration column, it will perform sorting on row.duration instead of row.taskMetrics.executorRunTime, thus, not getting the desired results.

Can you please verify whether sorting on Duration column works as expected and fix if necessary? Thank you.

SparkQA commented:
**[Test build #99542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99542/testReport)** for PR 23160 at commit [`71aff34`](https://github.com/apache/spark/commit/71aff3447b4c1a80e168a18202864191df189709).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99542/
Test PASSed.

SparkQA commented:
**[Test build #99543 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99543/testReport)** for PR 23160 at commit [`661566b`](https://github.com/apache/spark/commit/661566bd51efb40176327a58e5bed5dcfe7cacf0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99543/
Test PASSed.

shahidki31 committed:
address comment

shahidki31 commented:
> So as far as I can think, I expect that if you sort the Duration column, it will perform sorting on row.duration instead of row.taskMetrics.executorRunTime, thus, not getting the desired results.

@pgandhi999 Actually we have mapped the "Duration" to "executorRunTime" for sorting in the PR which I have mentioned above. So, after this PR the "Duration" is sorting correctly

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5610/
Test PASSed.

SparkQA commented:
**[Test build #99549 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99549/testReport)** for PR 23160 at commit [`5369a86`](https://github.com/apache/spark/commit/5369a8665ef71c44fb34ad68c5a600c7f8c3eeb4).

pgandhi999 commented:
LGTM

shahidki31 commented:
cc @tgravescs @srowen

srowen commented:
Merged to master. @shahidki31 does this need to go in branch 2.4, 2.3? I tried back porting it, but looks like a lot of the affected code didn't exist in 2.4. If the fix can or should also be back-ported and you're willing, you're welcome to open another PR against 2.4.

shahidki31 commented:
Thanks @srowen. This issue happens only in master branch.

Thank you all for the review and comments
